### ISSUE TITLE
TF2.0   'NoneType' object has no attribute 'shape'

### ISSUE THREAD

**Issue by zwenju** [tensorflow/tensorflow#31891]
windows 10
anaconda python 3.7 

**Describe the current behavior**
![image](https://user-images.githubusercontent.com/27112868/63515144-ba7cad80-c51c-11e9-93cb-167363519e4f.png)

Problem 
Line 32 input_x[:,1] occurs the error.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

```
import time
import math
import tensorflow as tf
import numpy as np
import tensorflow as keras
from tensorflow.keras import layers
from tensorflow import keras
import tensorflow.keras.backend as K


from tensorflow.keras.utils import plot_model
from scipy . stats import multivariate_normal as normal

tf.keras.backend.set_floatx('float64')


def test_tf (x):
    z = x[:,0]** 3 + x[:,1]**3
    return z


def grad (y, x):
    z = layers.Lambda ( lambda z : K.gradients(z[0], z[1]) ) ([y, x])
    return z

input_x  = keras.Input( shape = (2))
y =  test_tf (input_x)

print("y = ", y )
print ("input_x[0] = ", input_x[:1])

z0 = grad(y, input_x[:,1])

print("z0 = ", z0)

model =  keras.Model(inputs = [input_x], outputs = [z0] )

print ("gradient  = ", model.predict(  tf.constant( [[2,3]]  )   ))

```

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by gowthamkpr** [comment]
The error is caused in the line `z0 = grad(y, input_x[:,1])` as there is a shape mismatch
Plese find the [github gist](https://colab.research.google.com/gist/gowthamkpr/a1acdcf518d8930f81cde19acfe98ac3/gradients.ipynb) where I have solved this problem and let me know if it helps. Thanks!

**Comment by zwenju** [comment]
No, this  is not my expected solution.

I know  that   z0 = grad(y, input_x)  will make the code run.

What  wanted is that   we only want part of derivatives with input_x,
so  I write it as    z0 = grad(y, input_x[:. 0]).     

Could you solve this bug   without changing Z0 = grad(y, input_x[:,0]) ?

**Comment by WindQAQ** [comment]
Hi @zwenju, to my best knowledge, it's impossible to make your code run without changing `z0 = grad(y, input_x[:, 0])`.

1. If you use keras functional API, you have to make sure *all tensor operations* are done in `tf.keras.layers.Layer` or its subclass, including *slicing operation* as you use `input_x[:, 0]`.

2. Moreover, since *slicing operations* create a new tensor, if you try to compute gradients from the return value of `input[:, 0]`, you will get the value of `None` because of unconnected computational graph. The solution is to compute gradients w.r.t. `x`, and slice whatever you want from it. So the runnable example should be like this:

```python
import tensorflow as tf
import tensorflow.keras.backend as K
from tensorflow import keras
from tensorflow.keras import layers

tf.keras.backend.set_floatx('float64')


def test_tf(x):
    z = x[:, 0] ** 3 + x[:, 1]**3
    return z


input_x = keras.Input(shape=(2))
y = layers.Lambda(test_tf)(input_x)


def grad_fn(w):
    z = K.gradients(w[0], w[1])[0][:, 0]
    return z


z0 = layers.Lambda(grad_fn)([y, input_x])

model = keras.Model(inputs=[input_x], outputs=[z0])

print("gradient  = ", model.predict(tf.constant([[2, 3]])))
```

BTW, the example without functional API is in the following:
```python
import tensorflow as tf


def fn(x):
    return x[0] ** 2 + x[1] ** 2


x = tf.constant([1., 2.])

with tf.GradientTape() as tape:
    tape.watch(x)
    w = x[0]
    tape.watch(w)
    y = fn(x)

grads = tape.gradient(y, [x, w])
print(grads)
```
If you run it, you will see the gradients w.r.t. `x` is [2.0, 4.0] while the one w.r.t. `w` is `None`. Thanks.

**Comment by zwenju** [comment]
Thanks， 
In  issue  #31516，  similarly,  Input[:,0] creates a connected computational graph.
In this part,  it may exist a  non-consistency between **the slicing operation** and the the **gradient functions.**
Could you further consider it ???


Since slicing operations create a new tensor, if you try to compute gradients from the return value of input[:, 0], you will get the value of None because of unconnected computational graph.

**Comment by WindQAQ** [comment]
I do not really understand how two examples are related. If you plot your model in #31516, you can find that paths from prediction to `model.trainable_variables`, that is, batch normalization and dense layer (`BLOCK_*_{BN,DENSE}` in the following figure) does not go through any sliced tensor/op. Therefore the gradients are computed safely without obtaining any `None` value. However, if you try to compute gradient w.r.t. to `X` or `dW`, gradients will still be `None`.

![test](https://user-images.githubusercontent.com/11615393/63670387-15f5b680-c80f-11e9-98f4-0e41fdb6f9af.png)

**Comment by zwenju** [comment]
thanks WinQAQ, thanks for you solution.
Great!
